### ISSUE TITLE
So....Shrinking a database???? that is a bad idea

### ISSUE THREAD

**Issue by trouble2** [IowaComputerGurus/dnn.scheduledjobs#13]
Please read this: https://www.brentozar.com/archive/2009/08/stop-shrinking-your-database-files-seriously-now/ for reference
For some years though I had this idea of building something in combination with the world famous Ola Hallengren scripts.
Perhaps this would be something to look at, instead of using this...
https://ola.hallengren.com/

**Comment by trouble2** [comment]
I actually already run this SP (I found somewhere) each day on some DNN Installs to reindex or re-organize the index on tables with a lot of fragmentation:

CREATE PROCEDURE [dbo].[Bond_IndexDefrag] 
AS

DECLARE @COUNT_INDEXES INT = 1
DECLARE @TOTALTEMP_INDEXES INT
DECLARE @COUNT_TABLES INT = 1
DECLARE @TOTALTEMP_TABLES INT
DECLARE @QUERY VARCHAR (MAX) = ''
DECLARE @INDEXNAME VARCHAR (50)
DECLARE @TABLENAME VARCHAR (50)
DECLARE @FRAGMENTATION DECIMAL(18,2)
DECLARE @DBNAME VARCHAR (50) = DB_NAME()
DECLARE @NEWLINE AS CHAR(2) = CHAR(13) + CHAR(10)


	/* Log de gegevens voor en na de optimalisatie */

	--Select all tables and views in the db and add them to a temp. table
	SELECT 
		ROW_NUMBER() OVER (Order by TABLE_NAME) AS ID, TABLE_NAME
	INTO
		#TEMPTABLES
	FROM 
		INFORMATION_SCHEMA.TABLES
	WHERE 
		TABLE_TYPE IN ('BASE TABLE','VIEW') AND TABLE_CATALOG = @DBNAME

	SELECT 
		@TOTALTEMP_TABLES = COUNT(ID) 
	FROM 
		#TEMPTABLES --Total of the tables and views

	--check if there are any
	IF @TOTALTEMP_TABLES > 0
	BEGIN

		--while there are tables and views....
		WHILE @COUNT_TABLES <= @TOTALTEMP_TABLES
		BEGIN

			--Get the number of the table or view
			SELECT @TABLENAME = TABLE_NAME FROM #TEMPTABLES WHERE ID = @COUNT_TABLES

			/*
			only select tables and views with more then 5% fragementation and more then 1000 page_count
			*/
			SELECT 
				ROW_NUMBER() OVER (Order by avg_fragmentation_in_percent) AS ID, 
				Name, 
				avg_fragmentation_in_percent as Fragmentation
			INTO 
				#TEMPINDEXES 
			FROM
				sys.dm_db_index_physical_stats (DB_ID(), OBJECT_ID(@TABLENAME), NULL, NULL, NULL) AS a
			JOIN 
				sys.indexes AS b 
			ON
				a.object_id = b.object_id 
			AND 
				a.index_id = b.index_id
			WHERE 
				name IS NOT NULL
			AND 
				avg_fragmentation_in_percent > 5.00
			AND 
				page_count >= 500

			SELECT 
				@TOTALTEMP_INDEXES = COUNT(ID) 
			FROM 
				#TEMPINDEXES --Total of the indexes we will go trough

			--Check if there is anything to do
			IF @TOTALTEMP_INDEXES > 0
			BEGIN

				SET @COUNT_INDEXES = 1 --Reseteo contador de indices

				--loop trough the items
				WHILE @COUNT_INDEXES <= @TOTALTEMP_INDEXES
				BEGIN
					SELECT @INDEXNAME = Name FROM #TEMPINDEXES WHERE ID = @COUNT_INDEXES
					SELECT @FRAGMENTATION = Fragmentation FROM #TEMPINDEXES WHERE ID = @COUNT_INDEXES

					/*
					When the fragementation is less then 30%, we will do a reorganize, otherwise we will do a rebuild
					*/
					IF @FRAGMENTATION <= 30.00 BEGIN
						IF EXISTS (SELECT Name FROM sys.indexes WHERE name = @INDEXNAME AND object_id = OBJECT_ID(@TABLENAME))
						   BEGIN
								SET @QUERY += 'ALTER INDEX ' +  @INDEXNAME + ' ON ' + @TABLENAME + ' REORGANIZE'  + @NEWLINE
						   END
					END
					ELSE
					BEGIN
						IF EXISTS (SELECT Name FROM sys.indexes WHERE name = @INDEXNAME AND object_id = OBJECT_ID(@TABLENAME))
						   BEGIN
								SET @QUERY += 'ALTER INDEX ' +  @INDEXNAME + ' ON ' + @TABLENAME + ' REBUILD with (FILLFACTOR=85)' + @NEWLINE
						   END
					END

					SET @COUNT_INDEXES = @COUNT_INDEXES + 1
				END
			END

			DROP TABLE #TEMPINDEXES --remove the temp. table
			SET @COUNT_TABLES = @COUNT_TABLES + 1 
		END

		EXEC (@QUERY) --Execute the actual query(s)

		DROP TABLE #TEMPTABLES --Remove the other temp. table
		
	END


GO

**Comment by mitchelsellers** [comment]
I agree with the article, you shouldn't be doing a Shrink all of the time, but at the same time when we initially created that task and included it in the tooling the DNN Database would regularly grow in gigabites inappropriately and needed to shrink down.
